building py_sources fails

When building the py_sources stack (`./scripts/control_stack.sh py_sources build`) the build fails build the required dependency wheels:
This is the failing dockerfile line:
 https://github.com/NOAA-OWP/DMOD/blob/a355f272a17f14b839b7d28d43a4e2c9886ca694/docker/py_sources/py_sources.Dockerfile#L8
Giving this error:
```
 Building wheels for collected packages: cffi
    Building wheel for cffi (setup.py): started
    Building wheel for cffi (setup.py): finished with status 'error'
    ERROR: Command errored out with exit status 1:
     command: /usr/local/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-cqqauhp5/cffi/s
etup.py'"'"'; __file__='"'"'/tmp/pip-install-cqqauhp5/cffi/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);cod
e=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pi
p-wheel-p4kxp7xk --python-tag cp37
         cwd: /tmp/pip-install-cqqauhp5/cffi/
    Complete output (48 lines):
    unable to execute 'gcc': No such file or directory
```

If I add these lines
```
ARG REQUIRE="gcc"
RUN apk update && apk upgrade && apk add --no-cache ${REQUIRE}
```
to the beginning of the dockerfile to ensure gcc is available, I can get past this error, but then find myself facing missing object files:
```
Building wheels for collected packages: cffi
    Building wheel for cffi (setup.py): started
    Building wheel for cffi (setup.py): finished with status 'error'
    ERROR: Command errored out with exit status 1:
     command: /usr/local/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-bc6qm4rl/cffi/s
etup.py'"'"'; __file__='"'"'/tmp/pip-install-bc6qm4rl/cffi/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);cod
e=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pi
p-wheel-q3g0yioe --python-tag cp37
         cwd: /tmp/pip-install-bc6qm4rl/cffi/
    Complete output (44 lines):
    /usr/lib/gcc/x86_64-alpine-linux-musl/8.3.0/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find Scrt1.o: No such file o
r directory
    /usr/lib/gcc/x86_64-alpine-linux-musl/8.3.0/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find crti.o: No such file or
 directory
    /usr/lib/gcc/x86_64-alpine-linux-musl/8.3.0/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find -lssp_nonshared

    collect2: error: ld returned 1 exit status
```

This will finally leave you with:
```
gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -DTHREAD_STACK_SIZE=0x100000 -fPIC -DUSE__THREAD -DHAVE_
SYNC_SYNCHRONIZE -I/usr/include/ffi -I/usr/include/libffi -I/usr/local/include/python3.7m -c c/_cffi_backend.c -o build/temp.lin
ux-x86_64-3.7/c/_cffi_backend.o
    c/_cffi_backend.c:15:10: fatal error: ffi.h: No such file or directory
     #include <ffi.h>
              ^~~~~~~
    compilation terminated.
    error: command 'gcc' failed with exit status 1
```
which can be resolved by adding `libffi-dev` to the apk `${REQUIRE}` var.

I have a line some thing like this:
RUN apk update && apk add --no-cache gcc libc-dev openssl openssl-dev
python3 python3-dev libffi-dev bash && rm -rf /var/cache/apk/* && alias
python=python3 &
& pip3 install --no-cache-dir --upgrade pip

which install gcc and libraries


On Mon, Aug 31, 2020 at 3:24 PM Nels <notifications@github.com> wrote:

> If I add these lines
>
> ARG REQUIRE="gcc"
> RUN apk update && apk upgrade && apk add --no-cache ${REQUIRE}
>
> to the beginning of the dockerfile to ensure gcc is available, I can get
> past this error, but then find myself facing missing object files:
>
> Building wheels for collected packages: cffi
>     Building wheel for cffi (setup.py): started
>     Building wheel for cffi (setup.py): finished with status 'error'
>     ERROR: Command errored out with exit status 1:
>      command: /usr/local/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-bc6qm4rl/cffi/s
> etup.py'"'"'; __file__='"'"'/tmp/pip-install-bc6qm4rl/cffi/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);cod
> e=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pi
> p-wheel-q3g0yioe --python-tag cp37
>          cwd: /tmp/pip-install-bc6qm4rl/cffi/
>     Complete output (44 lines):
>     /usr/lib/gcc/x86_64-alpine-linux-musl/8.3.0/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find Scrt1.o: No such file o
> r directory
>     /usr/lib/gcc/x86_64-alpine-linux-musl/8.3.0/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find crti.o: No such file or
>  directory
>     /usr/lib/gcc/x86_64-alpine-linux-musl/8.3.0/../../../../x86_64-alpine-linux-musl/bin/ld: cannot find -lssp_nonshared
>
>     collect2: error: ld returned 1 exit status
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-OWP/DMOD/issues/20#issuecomment-684017906>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACA4SROI2XEUZY4SV4J7U7DSDQBHFANCNFSM4QQ4ZKDQ>
> .
>

--- a/docker/py_sources/py_sources.Dockerfile
+++ b/docker/py_sources/py_sources.Dockerfile
@@ -29,7 +29,7 @@ ARG externalrequests_package_name
 ARG scheduler_package_name
 # Set this so script below will not run logic that only applies to when in a full Git repo directory tree
 ENV OUT_OF_GIT_REPO=true
-RUN for p in communication access externalrequests scheduler; do \
+RUN for p in communication access externalrequests scheduler redis; do \
         ./scripts/dist_package.sh --sys python/lib/${p} && mv python/lib/${p}/dist/*.whl /DIST/.; \
     done
 ################################################################################################################

--- a/docker/py_sources/py_sources.Dockerfile
+++ b/docker/py_sources/py_sources.Dockerfile
@@ -2,6 +2,8 @@
 ################################################################################################################
 ##### Create base level intermediate build stage
 FROM python:3.7-alpine as basis
+ARG REQUIRE="gcc musl-dev libffi-dev"
+RUN apk update && apk upgrade && apk add --no-cache ${REQUIRE}
 # Copy project requirements file, which should have everything needed to build any package within project
 COPY ./requirements.txt /nwm_service/requirements.txt
 # Along with setup and wheel to build, install all project pip dependencies for package building later

